litebird_sim/distribute.py

```diff
@@ -0,0 +1,215 @@
+    a distribution that tries to assign the most uniform number of
+    elements to each group.
+
+    This function works even if ``num_of_elements < num_of_groups``.
+
+    .. doctest::
+
+      >>> distribute_evenly(5, 2)
+      [Span(start_idx=0, num_of_elements=3), Span(start_idx=3, num_of_elements=2)]
+
+      >>> distribute_evenly(1, 2)
+      [Span(start_idx=0, num_of_elements=1), Span(start_idx=1, num_of_elements=0)]
+
+    Args:
+        num_of_elements (int): The number of elements to distribute
+        num_of_groups (int): The number of groups to use in the result
+
+    Returns:
+        a list of 2-element tuples containing `num_of_groups`
+        elements, each of them being a 2-element tuple containing (1)
+        the index of the element in the group and (2) the number of
+        elements in the group. Being named tuples, you can access the
+        index using the field name ``start_idx``, and the number of
+        elements using the name ``num_of_elements``.
+
+    """
+    assert num_of_elements > 0
+    assert num_of_groups > 0
+
+    base_length = num_of_elements // num_of_groups
+    leftovers = num_of_elements % num_of_groups
+
+    # If leftovers == 0, then the number of elements is divided evenly
+    # by num_of_groups, and the solution is trivial. If it's not, then
+    # each of the "leftoverss" is placed in one of the first groups.
+    #
+    # Example: let's split 8 elements in 3 groups. In this case,
+    # base_length=2 and leftovers=2 (elements #7 and #8):
+    #
+    # +----+----+----+  +----+----+----+  +----+----+
+    # | #1 | #2 | #3 |  | #4 | #5 | #6 |  | #7 | #8 |
+    # +----+----+----+  +----+----+----+  +----+----+
+    #
+    #
+
+    result = []
+    for i in range(num_of_groups):
+        if i < leftovers:
+            # Make place for one of the leftovers
+            cur_length = base_length + 1
+            cur_pos = cur_length * i
+        else:
+            # No need to accomodate for leftovers, but consider their
+            # presence in fixing the starting position for this group
+            cur_length = base_length
+            cur_pos = base_length * i + leftovers
+
+        result.append(Span(start_idx=cur_pos, num_of_elements=cur_length))
+
+    assert len(result) == num_of_groups
+    assert sum([pair.num_of_elements for pair in result]) == num_of_elements
+    return result
+
+
+# The following implementation of the painter's partition problem is
+# heavily inspired by the code at
+# https://www.geeksforgeeks.org/painters-partition-problem-set-2/?ref=rp
+
+
+def _num_of_workers(arr, n, maxLen, weight_fn):
+    total = 0
+    num_of_workers = 1
+
+    for cur_element in arr:
+        cur_weight = weight_fn(cur_element)
+        total += cur_weight
+
+        if total > maxLen:
+            num_of_workers += 1
+            # Reset "total" for the next worker
+            total = cur_weight
+
+    return num_of_workers
+
+
+def _find_max_and_sum(arr, weight_fn):
```

Comment:
what about
```
weights = list(map(weight_fn, arr))
return (max(weights), sum(weights))
```

Reply:
Right, my first implementation used Numba and I decided to unroll the `for` loop to avoid iterating the same list twice. However, I realized that this function is not worth to be optimized, as it has to process very few elements. I'll apply your definition.

Reply:
Done.